src/lib.rs

```diff
@@ -68,19 +68,154 @@ use std::{
+    fn with_as_ffi_context<T, F: FnOnce(&mut FfiContext) -> T>(&mut self, closure: F) -> T;
+}
+
+impl<'a> ContextExt for Context<'a> {
+    fn with_as_ffi_context<T, F: FnOnce(&mut FfiContext) -> T>(&mut self, closure: F) -> T {
+        #[repr(C)]
+        struct FfiWakerImplOwned {
+            vtable: &'static FfiWakerVTable,
+            waker: Waker,
+        }
+
+        static C_WAKER_VTABLE_OWNED: FfiWakerVTable = {
+            unsafe extern "C" fn clone(data: *const FfiWaker) -> *const FfiWaker {
+                let waker: Waker = (*data.cast::<FfiWakerImplOwned>()).waker.clone();
+                Box::into_raw(Box::new(FfiWakerImplOwned {
+                    vtable: &C_WAKER_VTABLE_OWNED,
+                    waker,
+                }))
+                .cast()
+            }
+            // In this case, we must own `data`. This can only happen on the `CRawWaker` returned from `clone`.
+            // Thus the `data` is a `Box<Waker>`.
+            unsafe extern "C" fn wake(data: *const FfiWaker) {
+                let b = Box::from_raw(data as *mut FfiWakerImplOwned);
+                b.waker.wake();
+            }
+            unsafe extern "C" fn wake_by_ref(data: *const FfiWaker) {
+                (*data.cast::<FfiWakerImplOwned>()).waker.wake_by_ref();
+            }
+            // Same as `wake`.
+            unsafe extern "C" fn drop(data: *const FfiWaker) {
+                let b = Box::from_raw(data as *mut FfiWakerImplOwned);
+                std::mem::drop(b);
+            }
+            FfiWakerVTable {
+                clone,
+                wake,
+                wake_by_ref,
+                drop,
+            }
+        };
+
+        #[repr(C)]
+        struct FfiWakerImplRef {
+            vtable: &'static FfiWakerVTable,
+            waker: *const Waker,
+        }
+
+        static C_WAKER_VTABLE_REF: FfiWakerVTable = {
+            unsafe extern "C" fn clone(data: *const FfiWaker) -> *const FfiWaker {
+                let waker: Waker = (*(*data.cast::<FfiWakerImplRef>()).waker).clone();
+                Box::into_raw(Box::new(FfiWakerImplOwned {
+                    vtable: &C_WAKER_VTABLE_OWNED,
+                    waker,
+                }))
+                .cast()
+            }
+            unsafe extern "C" fn wake_by_ref(data: *const FfiWaker) {
+                (*(*data.cast::<FfiWakerImplRef>()).waker).wake_by_ref();
+            }
+            unsafe extern "C" fn unreachable(_: *const FfiWaker) {
+                abort();
+            }
+            FfiWakerVTable {
+                clone,
+                wake: unreachable,
+                wake_by_ref,
+                drop: unreachable,
+            }
+        };
+
+        let waker = FfiWakerImplRef {
+            vtable: &C_WAKER_VTABLE_REF,
+            waker: self.waker(),
+        };
+
+        let mut ctx = FfiContext {
+            waker_ref: unsafe { std::mem::transmute(&waker) },
```

Comment:
Bare transmute is not good. We should always specify type arguments using `::<>`.

Also if we have `waker_ref: *const FfiWaker` as I mentioned above, we just need `as` to cast pointers and no need for `transmute`.

Reply:
The reason why I changed it to a reference in the first place was that it makes the compiler force us to reference a valid `FfiWaker`. When it was a pointer, in my original implementation that did not work I made a `FfiContext` with a pointer to a `FfiWaker` and then dropped it without even noticing. If it was a reference like it is now, the compiler would have prevented me from doing that.